fix(Examples,PeriphDrivers): Avoid SIMO soft-start wakeup in LP example for MAX32665

victor-levi-maxim opened the pull request:
Added workaround to avoid SIMO soft start on wakeup form deepsleep and backup.

lorne-maxim commented on Examples/MAX32665/LP/main.c:
```diff
@@ -205,12 +205,65 @@ void prepForDeepSleep(void)
+
+void prepForBackup(void)
+{
+    MXC_ICC_Disable();
```
Should the prepForBackup and prepForDeepSleep functions be combined into one with an input parameter to select the mode?

victor-levi-maxim replied:
Got it. Will require a bit more work on that.

victor-levi-maxim replied:
Moved SIMO workaround code into PeriphDrivers. Still have separate prep functions for deep sleep and backup mode. I feel it is simpler and easier to use this way; not clear to me if we gain anything by combining into one function. There will be some overhead for passing a parameter and extra if/else inside the function.

lorne-maxim reviewed (changes requested):
I'd like to go ahead and move the prep and recover functions into the LP peripheral driver.  The MXC_LP_EnterDeepSleepMode and MXC_LP_EnterBackupMode should be handling the workaround automatically.

One thing to watch out for when moving these functions.  The recover functions always enable ICC on exit.  ICC should only be enabled if it was already enabled when the user called the enter deepsleep/backup functions.

victor-levi-maxim commented:
/clang-format-run

victor-levi-maxim commented:
/clang-format-run

victor-levi-maxim commented:
Added save/restore state of ICache and system clock for DeepSleep. Backup mode does not require save/restore as it exits to Reset_Handler(). Incorporated all known deepsleep, backup, and recover workarounds into Enter<LPmode>() functions in PeriphDrivers. Please review the changes.

lorne-maxim reviewed (changes requested):
It looks like some files not related to the softstart workaround got added to this PR.  I have have not reviewed those.  Can you confirm only the files you intended to change are part of the PR?

victor-levi-maxim commented:
/clang-format-run

victor-levi-maxim commented on Libraries/PeriphDrivers/Include/MAX32665/lp.h:
```diff
@@ -576,6 +576,25 @@ void MXC_LP_ICache0LightSleepEnable(void);
  */
 void MXC_LP_ICache1LightSleepEnable(void);
 
+/**
```
Removed unused declarations.

victor-levi-maxim commented:
/clang-format-run

sihyung-maxim commented on Libraries/PeriphDrivers/Source/LP/lp_me14.c:
```diff
@@ -574,3 +549,226 @@ void MXC_LP_ICacheXIPLightSleepEnable(void)
+ *
+ *  Enable the HIRC, set the divide ration to /4, and disable the 96 MHz oscillator.
+ */
+void switchToHIRCD4(void)
```
Should this be a static function as well?

victor-levi-maxim replied:
Yes, the more static functions the better.

sihyung-maxim commented on Libraries/PeriphDrivers/Source/LP/lp_me14.c:
```diff
@@ -574,3 +549,226 @@ void MXC_LP_ICacheXIPLightSleepEnable(void)
+            MXC_MCR->ctrl |= MXC_F_MCR_CTRL_VDDCSWEN;
+
+            /* SIMO clock setup for deep sleep */
+            *(volatile int *)0x40005434 = 1; /* SIMOCLKDIV [1:0] : 0=div1; 1=div8; 2=div1 3=div16 */
+            /* BUCK_CLKSEL [25:24] : 0=8K; 1=16K; 2=30K; 3=RFU */
+            *(volatile int *)0x40005440 = (*(volatile int *)0x40005440 & (~(0x3 << 24))) |
+                                          (0x2 << 24);
+            /* BUCK_CLKSEL_LP [7:6] : 0=8K; 1=16K; 2=30K; 3=RFU */
+            *(volatile int *)0x40005444 = (*(volatile int *)0x40005444 & (~(0x3 << 6))) |
+                                          (0x2 << 6);
```
Should these register locations be brought out to the register files? We can add them to pwrseq_regs.h/pwrseq_me14.svd if these are missing registers that should be included.

victor-levi-maxim replied:
Not sure. I have no objections to adding registers. However there must be a reason why those registers were not added in the first place. I suspect those are "hidden" registers that customers should not mess with.

sihyung-maxim replied:
Okay, I'll check with Hiep.

sihyung-maxim commented on Libraries/PeriphDrivers/Source/LP/lp_me14.c:
```diff
@@ -574,3 +549,226 @@ void MXC_LP_ICacheXIPLightSleepEnable(void)
+    /* Go into Deepsleep mode and wait for an interrupt to wake the processor */
+    __WFI();
+
+    /*void recoverFromDeepSleep(void)*/
+    {
+        /* SIMO soft start workaround on wakeup */
+        /*void MXC_LP_recoverFromDeepSleep(void)*/
+        {
```
Not sure if I'm missing something or if I accidentally deleted some files while reverting the clang-format, but there are hanging curly braces in lp_me14.c.

victor-levi-maxim replied:
Braces should be balanced. Some extra brace pairs are just leftovers from copy-paste a function. Function name commented out, but braces were left to show that function limits. I left extra braces in since they do no harm. Can be removed.

sihyung-maxim replied:
Got it. I think it'd be best to remove the curly braces and the unnecessary commented-out code, so it's more readable for users.

victor-levi-maxim replied:
> Got it. I think it'd be best to remove the curly braces and the unnecessary commented-out code, so it's more readable for users.

Done.

sihyung-maxim reviewed (changes requested):
Please review my reverted and merged commits. Hopefully, nothing relating to the LP changes was deleted when I was cleaning up the clang-format.

victor-levi-maxim commented:
> Please review my reverted and merged commits. Hopefully, nothing relating to the LP changes was deleted when I was cleaning up the clang-format.

Look good. I added static for switchToHIRCD4().

sihyung-maxim commented:
/clang-format-run
